server/webapp/WEB-INF/rails.new/webpack/views/preferences/email_settings.js

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+;(function() { // eslint-disable-line no-extra-semi
+  "use strict";
+
+  const m = require("mithril");
+  const Stream = require("mithril/stream");
+
+  function EmailSettings(model, smtpEnabled=false) {
+    const editMode = Stream(false);
+    const allowCheckbox = editMode.map((value) => smtpEnabled && value);
+
+    function enterEditMode(e) {
+      if (e) {
+        e.preventDefault();
+      }
+
+      editMode(true);
+      model.reset();
+    }
+
+    function exitEditMode(e) {
+      if (e) {
+        e.preventDefault();
+      }
+
+      editMode(false);
+      model.reset();
+    }
+
+    return {
+      oninit() {
+        model.load();
+      },
+
+      view() {
+        return m("form", model.config({class: "email-settings"}, exitEditMode),
+          m("legend", "Email Settings"),
+          m("fieldset", [
+            m(LockableInput, {name: "email", label: "Email", value: model.email, unlocked: editMode, placeholder: "Email not set"}),
+            m(LockableCheckbox, {name: "email_me", label: "Enable email notification", value: model.enableNotifications, unlocked: allowCheckbox}),
+            m(LockableInput, {name: "checkin_aliases", label: "My check-in aliases", value: model.checkinAliases, unlocked: editMode, placeholder: "No matchers defined"}),
+          ]),
+          m("fieldset",
+            editMode() ? [
+              m("input", {type: "submit", value: "Save", class: "primary"}),
+              m("input", {type: "reset", value: "Cancel", onclick: exitEditMode})
+            ] : m("input", {type: "button", value: "Edit", onclick: enterEditMode})
+          )
+        );
+      }
+    };
+  }
+
+  const LockableCheckbox = {
+    oninit(vnode) {
+      vnode.state.name = vnode.attrs.name;
+      vnode.state.value = vnode.attrs.value;
+      vnode.state.label = vnode.attrs.label;
+      vnode.state.unlocked = vnode.attrs.unlocked;
```

Comment:
Do not store the state in a component.
`oninit` method is called only while creating the html element. And when the redraw happens, only the `view` of a component is recomputed.
In above scenarios, if the data changes after the intializing the component, it will end up showing stale information.

Instead in the views, we could use:
```
<input name={vnode.attrs.name} ... />
```